rareview/rareview-wordpress-coding-challenge-starter
# Add Books post type with supporting meta, templates, styles, and block

# What does this pull request introduce?

### There are 2 major items that are introduced in this pull request.

- A Books post type
- A Books block

## The Books Post Type

The Books post type features core support for basic components like title, editor, thumbnail, and excerpt, as well as a custom meta field for Book Author.

Accompanying the post type is a base `single-book.php` template that includes a sidebar for displaying the Book's thumbnail and a content area to display the title, author name, and post content.

## The Books Block

The Books block follows the required specs:

- Block that will be available only for the “post” post type.
- Block will allow to pick from a select box “Books” posts.
- Block will allow the chosen books to be re-ordered via drag and drop.
- On a “post” single page the Books block will render the list of books selected. They will show the title, excerpt, featured image and author name.

For a better user experience, I have also enhanced the functionality of the Books block to:

- Move the selection and reordering of Books into the InspectorControls
- Add a new component to the Editor to mimic the frontend display that updates as selections are added or removed.

# Implementation

### The Books post type and Author Name meta

To add the Books post type, I followed the structure found in the `wp-content/themes/rareview/includes/classes/PostTypes` directory and included the new post type in `wp-content/themes/rareview/includes/classes/PostTypes/PostTypeServiceProvider.php`

For the Author Name meta, I included new constants in `wp-content/themes/rareview/assets/js/shared/constant.js` for `BOOK_POST_TYPE` and `BOOK_CPT_AUTHOR_META_KEY`. Using these constants, I followed the meta registration found in the `wp-content/themes/rareview/assets/js/admin/gutenberg/custom-sidebars/` directory and included the new custom-sidebar in `wp-content/themes/rareview/assets/js/admin/gutenberg/custom-sidebars/index.js`.

### The Books Block

To add the Books Block, I began by using the Example Block found in `wp-content/themes/rareview/includes/blocks/example-block` as the starting point.

I had begun with the most simple approach to see things working as expected and iterated from there until the final deliverable.

**Step 1 of iteration**
- Use WordPress Core's SelectControl allowing multiple selections
- Populate the options with Book posts returned from apiFetch
- Simple frontend output to ensure all the pieces connect correctly

**Step 2 of iteration**
- Replace SelectControl with 10up's ContentPicker component
  - This component achieves a better UX Book selection and also includes the drag-and-drop reordering.
- Include `selectedBooks` as a block attribute
  - This passed to the frontend template and was used with `wp_list_pluck` to extract the post IDs for the WP_Query `post__in` and `orderby`

**Step 3 of iteration**
- Limit this block to only be selectable in the context of WP Core's "Post" post type
  - A deregistration script is conditionally enqueued if the current post type is not post.

**Step 4 of iteration**
- Improve Editor UX
  - Moved the ContentPicker component in an InspectorControl
  - Add `selectedBookIds` to block attributes
  - Using `Array.map`, loop the `selectedBookIds` in an apiFetch to display the selected Books as they would be seen on the frontend
  - Update `markup.php` to use `selectedBookIds` vs plucking the IDs from `selectedBooks`

**Step 5 of iteration**
- Improve UI
  - Conditionally apply grid classes based on the number books selected
    - 4 books display in a 4 column grid
    - 3 books display in a 3 column grid
    - 2 books display centered in a 4 column grid
    - 1 book displays centered in a 3 column grid
  - All grid variations respond nicely based on the viewport width.

# Testing

### To test these additions, first set yourself up to test these updates in a local environment.
- Pull from https://github.com/maxinacube/rareview-wordpress-coding-challenge-starter
   - Switch to the `feature/book-post-type` branch
   - **Alternatively**, a feature branch may be created in the Rareview source repo and this PR may be merged there for testing.
- Once the `feature/book-post-type` branch has been pulled, do one of the following:
   - If you are already running this repo in a local environment, navigate to the root of the repo and run `npm run build`
      - Log-in to the WordPress admin, navigate to Settings > Permalinks and re-save the permalinks.
   - If you are not running this repo in a local environment, see [Local Development Setup](https://github.com/rareview/rareview-wordpress-coding-challenge-starter/blob/master/.lando/docs/local-development-setup.md) for details setup.
      - Be sure to run an `npm run build` to ensure you have access to the new Books block

### After running a build on the latest additions, it's time to test the new features.

**To test the Books post type**
1. Navigate to the Books custom post type in the WordPress admin, and click "Add New"
2. When presented with the edit screen, include the following data:
   - Title (Editor)
   - Book Author (Book sidebar)
   - Featured Image (Book sidebar)
   - Excerpt (Book sidebar)
3. Hit the "Publish" button

Repeat these steps for each additional Book. At this point, the single Book template can be viewed on the frontend by following the URL in the Book sidebar.

**To test the Books Block**
1. Navigate to the Posts posts type in the WordPress admin, and click "Add New" or edit an existing post.
2. Add the "Books" block to the post by
  - Entering to a new line and type `/Books` or
  - Use the Block Inserter (blue button with "+" icon) to select or search for "Books"
3. In the Block Sidebar, start typing in the "Search For Books" input to select which Books should appear.
4. Publish or Update your post
5. View the post by clicking the "View Post" notice, or by following the URL in the Post sidebar

# Future Improvements

- Add a `View Type` SelectControl to the Books Block Inspector Control
  - Default to "Grid" with option to select "List"
  - Apply `grid` class or `list` class conditionally to the Books container.
  - Pass the View Type to the Books components

- If View Type is Grid, include SelectControl for the user to select how many columns the grid is.
  - I am testing with an assumption that this is being viewed in a full-width view, but the Books block could be added in a 50/50 column leaving the display smushed with 3 or 4 books, or feeling out-of-place with 1 or 2 books.

- Pull the looped book item structure out of markup, place into a theme partial, and include that into the frontend block markup.
  - This could be beneficial if we wanted a Books Archive to use the same item layout that the Books Block provides.
